### ISSUE TITLE
Given a query, return a ranking

### ISSUE THREAD

**Issue by jmdu99** [rueycheng/AdaRank#4]
Hi @rueycheng,

After doing some research I think `print_trec_run` is not what I'm looking for. Perhaps you can help me with this:

Once I have the AdaRank model trained with a set of documents and queries in Svmlight format, I would like to return a ranking based on a query entered by the user. That is, I am not interested in generating a ranking from a test set with other different documents, but rather in having a ranking with scores for the documents with which I have trained the model. I don't know if I explain myself.

That is, after training the model:
```
 X, y, qid = load_svmlight_file(svm_file, query_id=True)
 model = AdaRank(max_iter=100, estop=10, scorer=NDCGScorer(k=10)).fit(X, y, qid)
```

I want to calculate the scores for the same `X` documents given a `qid_test` (or more than one). And with those scores, return a ranking like:
```
qid_test docno   rank   score
----------------------------
1        3       1      0.7
1        5       2      0.23
1        6       3      0
2        6       1      1
2        5       2      0
2        3       3      0
```

Is it possible to get something like this?

Thanks in advance.

**Comment by rueycheng** [comment]
Have a look at `test.py` in the repo. There's a place where you get `score`, which is essentially an array of scores, out of an NDCG scorer. I think that's what you're after.

Cheers

**Comment by jmdu99** [comment]
I already checked it. The problem I have is that I don't know how `X_test`, `y_test`, `qid_test` should look like. 

For example, if I have `X_train` (3 documents) like:
  ```
  (0, 12)	0.4217826840250329
  (0, 51)	0.4217826840250329
  (0, 68)	0.1512472665969307
  (0, 71)	0.320779238063236
  (0, 77)	0.4217826840250329
  (0, 81)	0.1512472665969307
  (0, 83)	0.373827070913866
  (0, 91)	0.4217826840250329
  (1, 20)	0.5645782010909166
  (1, 94)	0.8253795822849901
  (2, 35)	0.5773502691896257
  (2, 52)	0.5773502691896257
  (2, 58)	0.5773502691896257
```

`qid_train` (1 query) like: `[1 1 1]` and `y_train` (relevance of the 3 documents for that query) like: `[0. 0. 0.]`

Now a user introduces a query that is not in `qid_train`, but the documents are the same:

- Should  `qid_test` have a different identifier from `qid_train`? (for example: `[2 2 2]`)
- Should `X_test`  be exactly the same as `X_train`? (since the documents are exactly the same)
- Should `y_test` have the relevance scores for the new query? (for example: `[1 1 1]`)

And another 2 questions I have:

- How should the `docnos` be for documents that are the same but are accompanied by a different qid?:
```
0 qid:1 20:0.5645782010909166 94:0.8253795822849901 #2
1 qid:2 20:0.5645782010909166 94:0.8253795822849901 #2
```
That is, the `docno` must be the same or must be a different one? Is this OK?

- I have assigned the **relevance** between query and documents as follows:

  0 ~ no match between document and query
  1 ~ **partial** match between document and query
  2 ~ **exact** match between document and query

  Is this a correct way of assigning relevance for the NDCG scorer or can it be done better?

Thanks in advance

**Comment by rueycheng** [comment]
Hey @jmdu99, 

These questions are not relevant to AdaRank so posting them up as issues here is really not ideal. Can I refer you to some tutorial and resource pages as I think you'll find better answers there?
- https://www2009.thewebconf.org/pdf/T7A-LEARNING%20TO%20RANK%20TUTORIAL.pdf
- http://didawiki.di.unipi.it/lib/exe/fetch.php/magistraleinformatica/ir/ir13/1_-_learning_to_rank.pdf
- https://www.microsoft.com/en-us/research/project/letor-learning-rank-information-retrieval/

Regarding your setup, a quick take from me is that you should _separate test set entirely from training set_. Otherwise we can't be sure if the ranking model does learn to generalize across different queries/datasets. If you follow a proper train/test split and end up with two sets of query-doc sets, you naturally will have different qids, features, and labels and that would probably solve all your problems. 

You have the leisure to develop your own relevance levels and I'm pretty sure I've seen similar designs in the past - so it should be all good.

I'll close this issue for now. Best of luck hacking away on your own. Should you have any further question, please drop me a line via email.

Cheers,
RC
